### ISSUE TITLE
Incorrect result for BETWEEN query that casts column to boolean

### ISSUE THREAD

**Issue by mrigger** [cwida/duckdb#571]
Consider the following statements:
```sql
CREATE TABLE t0(c0 INT);
INSERT INTO t0 VALUES (-1);
SELECT t0.c0 FROM t0 WHERE NOT (0 BETWEEN 0 AND t0.c0::BOOL); -- expected: {}, actual: {-1}
```
Unexpectedly, the `SELECT` fetches a row, although the predicate should evaluate to `FALSE`. I found this bug based on commit a09d2f484db83ac9e27c9501f439bcd32b0083ec.

**Comment by Mytherin** [comment]
Apologies for taking some time to look at this one :)

The problem was that we ended up with a filter in the form of `0 <= t0.c0::BOOL::INTEGER`, and an optimizer then trigger to simplify this filter by moving the casts to the other side. So we got `0::BOOL::INTEGER <= t0.c0`.

This optimization only works on casts that are invertible, and resulted in a bug here because `INTEGER -> BOOLEAN -> INTEGER` does not necessarily produce the same integer again. In this case the casts went from `-1 -> TRUE -> 1`. 

I have now fixed this by adding a `CastIsInvertible` function that is checked before applying this optimization. That fixes the problem here.
